Add operators status to Dashboards

rawagner opened the pull request:
depends on https://github.com/openshift/console/pull/3754

![operators_status](https://user-images.githubusercontent.com/2078045/70635379-15edbb00-1c34-11ea-9d04-7a8f9d768059.png)

![operators_status_deg](https://user-images.githubusercontent.com/2078045/70635397-1dad5f80-1c34-11ea-80d2-69c998981424.png)

andybraren commented on frontend/packages/console-app/src/components/dashboards-page/OperatorStatus.tsx:
```diff
@@ -0,0 +1,228 @@
+}) => (
+  <>
+    <div className="co-operator-section">
+      Operators extend Kubernetes with additional custom resources to manage applications
```
```suggestion
      Operators extend OpenShift with additional custom resources to manage applications.
```

Added a period and switched to "OpenShift". We might want to validate the language here with an SME.

itsptk replied:
@dmesser @tlwu2013 I spoke with @alecmerdler and he mentioned we are transforming this CSV status in the Installed Operator details to show as just "Succeeded." Could that be appropriate here as well? This is a roll-up status of all the CSVs' statuses for an operator across namespaces, showing the "worst" status.

rawagner replied:
fixed

rawagner commented:
Updated status to read from `status.phase` instead of `status.reason`
![operators_succeeded](https://user-images.githubusercontent.com/2078045/70697998-e1770f00-1cc6-11ea-8b69-e67eb5a7f2a3.png)

![operators_installed_succeeded](https://user-images.githubusercontent.com/2078045/70698009-e5a32c80-1cc6-11ea-8f59-322c73b6da4e.png)

itsptk commented:
> Updated operator status to read from `status.phase` instead of `status.reason`

I think my inclination would be to continue to gather the operator status from reason and just transform the string in the status card (similar to the installed operator details.) Changing where this status is gathered from is a bigger change and would need more consideration.

rawagner commented:
> > Updated operator status to read from `status.phase` instead of `status.reason`
> 
> I think my inclination would be to continue to gather the operator status from reason and just transform the string in the status card (similar to the installed operator details.) Changing where this status is gathered from is a bigger change and would need more consideration.

CSV details page is using `status.phase` https://github.com/openshift/console/blob/master/frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx#L714 and CSV list page `status.reason`. So which one is correct/better ?

andybraren commented:
> CSV details page is using `status.phase` https://github.com/openshift/console/blob/master/frontend/packages/operator-lifecycle-manager/src/components/clusterserviceversion.tsx#L714 and CSV list page `status.reason`. So which one is correct/better ?

@dmesser @tlwu2013?

andybraren commented:
For the "Degraded" and "Upgrading" sublabels, would we be able to show a count of how many are degraded/upgrading like `2 Degraded` or `7 Upgrading`?

andybraren commented:
@rawagner I shared the original design at a UX meeting earlier today and got some great feedback about the all-okay state. I haven't gotten feedback on this design tweak yet, but here it is for reference:

![concept-operator-status](https://user-images.githubusercontent.com/9122899/70757517-e317f580-1d0d-11ea-8bbc-1c85c5c8ccc6.png)

Instead of showing the first 5 alphabetical operators that are all "Succeeded" in each section (which is somewhat arbitrary, distracting, and doesn't confirm that all the other operators are healthy) we'd show a collapsed "All succeeded" (or whatever it ends up being) status with the same "View all" link on the left.

Sorry for the mid-PR change, just trying to keep up with you. 🙂

itsptk commented:
Spoke with @tlwu2013 , we agreed that this popover should be using status.phase, since thats what the user would see in the the operator details if they clicked the operator name in popover. We didn't want to change the operator list view status at this time, as that is being actively reworked in https://jira.coreos.com/browse/PD-275

itsptk commented:
Just to be clear, if there is an operator that is in error in one namespace, clicking the name will take the user to the details of that operator, but if there is an operator in error in >1 namespaces, clicking the name will take the user to the list view of operators filtered? This interaction could change when the operators list changes to only show one row per operator when project selector is 'all projects' but I think it can be accommodated.

Also, will the status now be "All [status.phase]" when all are green?

andybraren commented:
> Spoke with @tlwu2013 , we agreed that this popover should be using status.phase, since thats what the user would see in the the operator details if they clicked the operator name in popover.

Sounds good, thanks for checking on that! 👍

> Just to be clear, if there is an operator that is in error in one namespace, clicking the name will take the user to the details of that operator, but if there is an operator in error in >1 namespaces, clicking the name will take the user to the list view of operators filtered? This interaction could change when the operators list changes to only show one row per operator when project selector is 'all projects' but I think it can be accommodated.

That behavior sounds correct to me for the way we represent operators in the UI today. When the [new cluster-scoped Operator CR](https://github.com/openshift/enhancements/blob/master/enhancements/olm/simplify-apis.md#proposal) arrives I imagine we'll always have a single Operator Details page that we can link to instead of sometimes linking to the List page. That should be nicer.

> Also, will the status now be "All [status.phase]" when all are green?

That's what I would propose for now. I assume the new cluster-scoped Operator CR would also have an aggregated `Status` of some kind, right? Would it make sense to show the number of degraded CSVs (like `3 degraded`) in this popover whenever clicking the name will take the user to a filtered List page?

I guess regardless of what we do in this PR, the way these statuses are shown should be revisited once the operator UI changes arrive to make sure they align.

rawagner commented:
Updates:
1. Added text under operators health item like `1 Failed`

![op_status_2](https://user-images.githubusercontent.com/2078045/70912063-72c1ea80-2013-11ea-8e89-6ae7b292a8f2.png)

The question I have here is whether the text should reflect the 'groupped' CVO operators (by name) - ie when we have operator running in 5 namespaces, 3 of those operators failed, should we say `1 Failed` or `3 Failed` ?
In Operator Status popup we have `Operators (2)` - this shows number of grouped operators - in this case we have `Local Storage` running in all namespaces and another operator running in one namespace - we show `Operators (2)` or should we show real number of operators (not grouped) ?

2. When one OLM operator failed the link will take you to it's details page, if more then link points to Installed Operators page filtered by name

3. Added `All [OK]` status
![op_status_1](https://user-images.githubusercontent.com/2078045/70912120-92f1a980-2013-11ea-9a65-e407ca3c7431.png)

note that CVO says `All succeeded` and CO says `All available` as the statuses for these resources are different and this way it aligns with their list pages

4. There's also discrepancy between not-ok states for CVO and CO - OLM says `Failed` with error icon in it's list page and CO says `Degraded` with Warning icon. Therefore if we have CVO in Failed state and CO in Degraded state - (error is worse than warning) the Operators Health will show error state and `1 Failed` (degraded CO is not counted since its only in warning state - but CO doesnt have worse status than warning)

rawagner commented:
> Spoke with @tlwu2013 , we agreed that this popover should be using status.phase, since thats what the user would see in the the operator details if they clicked the operator name in popover. We didn't want to change the operator list view status at this time, as that is being actively reworked in https://jira.coreos.com/browse/PD-275

reverted the change regarding status in CSV list page

andybraren commented:
@itsptk @tlwu2013 and I talked through this a bit and agree on the main pieces of the design below. It aligns with where we think a revised Installed Operators List page could go in the future, but we might need to revisit this card once that page is updated (it will probably group multiple operator instance statuses into 1 as well, with a status popover showing which projects have issues).

![concept-operator-status](https://user-images.githubusercontent.com/9122899/70946721-b291cd80-2025-11ea-999c-f1c132f309a3.png)

Note: The sentence describing operators has been updated, but we could still use an SME to double-check the wording for accuracy. I'm sure we'll start hearing from some when this merges. 🙂

### Status bar

The yellow/red icon is determined by the highest severity icon that the user would see when the popover is opened. For example, if the popover has 1 red error icon and 4 yellow warnings, the red error icon would appear in the status bar.

The count (i.e. `4 degraded`) is the total number of both red errors and yellow warning statuses that appear in the popover. The word `degraded` is always used, no matter what the actual labels in the popover are (hopefully the OLM operators and Cluster operators can agree on standardized status labels in the future).

### Status popover

The small `(3 installed)` text represents the number of operator groups there are by name. So if `Local Storage` is running in 17 projects, it's still counted as `(1 installed)` here.

The number of failed/degraded projects within that operator group is shown on the right, like `12 projects failed`.

Note: I'm not very familiar with the possible statuses for OLM operators. If they can be a mix of errors/warnings and various labels too, maybe `3 projects degraded` and using the icon of the highest severity would make sense to do here as well, similar to the Status bar's behavior described above.

When only 1 project is degraded, clicking the OLM operator name takes the user to the operator's Details page. When multiple projects are degraded, the user is taken to the List page filtered by that operator's name instead (hopefully we'll be able to filter by status in the future too).

andybraren commented:
> 4. There's also discrepancy between not-ok states for CVS and CO - CVS says `Failed` with error icon in it's list page and CO says `Degraded` with Warning icon.... CO doesnt have worse status than warning)

I see the `ClusterOperatorDegraded` alerting rule has a critical severity, so their status icons being yellow warnings and never red errors seems very weird. I'm not quite sure which team to ping about this, but it seems worth investigating. @benjaminapetersen? 🙂

andybraren commented:
Another thing that came up during discussions:

Is the Cluster Overview dashboard as a whole (including this new Operator status) only visible to users who have permission to view OLM & Cluster Operators (admins)? I'm not super familiar with how RBAC works, but we should make sure that only users with view access to OLM/CO operators would be able to see this new Operators status.

I'm not sure if we've done a full review/discussion about how RBAC works with our dashboards and which users can see what parts. Maybe we should investigate that @jelkosz @rawagner if we haven't already.

rawagner commented:
> Another thing that came up during discussions:
> 
> Is the Cluster Overview dashboard as a whole (including this new Operator status) only visible to users who have permission to view OLM & Cluster Operators (admins)? I'm not super familiar with how RBAC works, but we should make sure that only users with view access to OLM/CO operators would be able to see this new Operators status.
> 
> I'm not sure if we've done a full review/discussion about how RBAC works with our dashboards and which users can see what parts. Maybe we should investigate that @jelkosz @rawagner if we haven't already.

whole dashboards page is visible only to admins

rawagner commented:
@andybraren updated per your suggestions
![op_status_1](https://user-images.githubusercontent.com/2078045/71816244-42a1c080-3082-11ea-8e7e-05a2745fade3.png)
![op_status](https://user-images.githubusercontent.com/2078045/71816247-45041a80-3082-11ea-9b06-0e61997ab1b8.png)

andybraren commented:
Awesome, LGTM! FYI @itsptk @tlwu2013

vojtechszocs commented on frontend/packages/operator-lifecycle-manager/src/types.ts:
```diff
@@ -53,6 +53,12 @@ export enum SubscriptionState {
 }
 
+export enum ClusterServiceVersionStatus {
+  ClusterServiceVersionStatusFailed = 'Failed',
```
Nit: why the `ClusterServiceVersionStatus` prefix for all enum members?

rawagner replied:
the other enums in this file/package use the same naming so I went with what is already used to stay consistent

vojtechszocs commented on frontend/public/components/utils/async.tsx:
```diff
@@ -67,12 +67,16 @@ export class AsyncComponent extends React.Component<AsyncComponentProps, AsyncCo
+export type AsyncComponentProps = {
+  loader: () => Promise<React.ComponentType>;
+  LoadingComponent?: React.ReactNode;
+} & any;
```
FYI, the `& any` basically broadens the type declaration to be assignable to anything.

```ts
type Foo = { x: string };
const f1: Foo = { x: 'foo' };
const f2: (Foo & any) = true; // breaks the structural typing contract
```

I realize it's part of the original code, but it looks fishy..

vojtechszocs replied:
We can improve that later, but I'd encourage not to use `type Foo = xxx & any` in new code, just like with `as any` type cast.

rawagner commented on frontend/packages/console-app/src/extensions/dashboards.ts:
```diff
@@ -0,0 +1,63 @@
+import { K8sResourceCommon, K8sKind } from '@console/internal/module/k8s';
+import { HealthState } from '@console/shared/src/components/dashboard/status-card/states';
+
+namespace ExtensionProperties {
```
fixed, i moved the extension point directly to `console-plugin-sdk`

rawagner replied:
splitting the dashboards extensions to smaller pieces is a good idea, I will work on that in a followup to not make this PR any bigger

rawagner commented on frontend/packages/console-app/src/extensions/dashboards.ts:
```diff
@@ -0,0 +1,63 @@
+import { HealthState } from '@console/shared/src/components/dashboard/status-card/states';
+
+namespace ExtensionProperties {
+  export interface DashboardsOperatorStatus<R extends K8sResourceCommon> {
```
Added fallback to `K8sResourceCommon`
In case of URL subsystem, the `R` represents the response of a call to the given URL which can be anything like JSON or string, so i think that no fallback here is ok

vojtechszocs replied:
Right, I forgot that URL subsystem response can be anything.

rawagner commented on frontend/packages/console-app/src/components/dashboards-page/OperatorStatus.tsx:
```diff
@@ -0,0 +1,260 @@
+const OperatorsStatus = withDashboardResources(
+  ({ resources, watchK8sResource, stopWatchK8sResource }) => {
+    const additionalOperators = React.useMemo(
+      () => plugins.registry.get<DashboardsOperatorStatus<any>>(isDashboardsOperatorStatus),
```
not needed anymore as I moved the extension declaration to `console-plugin-sdk`, so fixed

rawagner commented on frontend/packages/console-app/src/components/dashboards-page/OperatorStatus.tsx:
```diff
@@ -0,0 +1,260 @@
+
+type OperatorStatusPopupProps = {
+  resources: FirehoseResourcesResult;
+  additionalOperators?: DashboardsOperatorStatus<any>[];
```
fixed

rawagner commented on frontend/packages/console-app/src/components/dashboards-page/operator-status.scss:
```diff
@@ -0,0 +1,47 @@
+@import '../../../../console-shared/src/styles/skeleton-screen.scss';
```
i remember trying the first approach before but it didnt work. But this is not needed anymore as I moved some files to `console-shared` together with this `scss`

vojtechszocs replied:
Yup, CSS styles should be close to related code if possible.

rawagner commented on frontend/packages/console-app/src/components/dashboards-page/utils.ts:
```diff
@@ -0,0 +1,43 @@
+} from '@console/shared/src/components/dashboard/status-card/states';
+
+export const OperatorHealthPriority: {
+  [key in keyof typeof healthStateMapping]: {
```
separate type was actually already defined so I reused it here

not it looks like `key in HealthStateMappingKeys`

rawagner commented on frontend/packages/console-app/src/components/dashboards-page/OperatorStatus.tsx:
```diff
@@ -0,0 +1,260 @@
+      ),
+    );
+
+    const PopupComponentCallback = React.useCallback(
```
no, I dont want the result (`useMemo`) of the call but I want to memoize the call and defer the rendering to later thus using `useCallback`

vojtechszocs commented on frontend/packages/operator-lifecycle-manager/src/plugin.tsx:
```diff
@@ -263,6 +268,29 @@ const plugin: Plugin<ConsumedExtensions> = [
+      ],
+      getOperatorsWithStatuses: getClusterServiceVersionsWithStatuses,
+      operatorRowLoader: () =>
+        import(
```
Let's stay consistent in how we use dynamic imports within this specific module:

```ts
loader: async () =>
  (await import('./components/dashboard/csv-status' /* webpackChunkName: "csv-dashboard-status" */))
    .default,
```

rawagner replied:
fixed

vojtechszocs commented on frontend/packages/operator-lifecycle-manager/src/status/csv-status.ts:
```diff
@@ -0,0 +1,62 @@
+    status: {
+      installedCSV: getName(csv),
+    },
+  } as any); // 'as any' to supress typescript error caused by lodash;
```
Can't we use generic type params when calling `_.find` function? Using `as any` is discouraged.

vojtechszocs commented on frontend/public/components/dashboard/dashboards-page/overview-dashboard/health-item.tsx:
```diff
@@ -0,0 +1,246 @@
+        />
+      );
+    })
+    .reverse();
```
What's the intention behind reversing the `sections` array?

rawagner replied:
we want to have CSV operators before Cluster operators thus we need to reverse the order of extensions (cluster operators are contributed by console app, CSV by operator-lifecycle). We can improve that later on if we will have any more complex cases, like including some extensions before/after others (similar to nav items) but reverse is enough for now

vojtechszocs replied:
Isn't it better then to explicitly sort the extensions in a way that Cluster operators come before any other ones?

vojtechszocs replied:
A better solution IMO is to add e.g. `priority` property to allow sorting extensions in a specific way when they are consumed.

For now, I'm OK with a simple `reverse`.

rawagner commented on frontend/packages/console-plugin-sdk/src/typings/dashboards.ts:
```diff
@@ -77,6 +81,27 @@ namespace ExtensionProperties {
+
+    /**
+     * Link to all resources page.
+     * If not provided than list page of first resource from resources prop is used.
```
fixed

rawagner commented on frontend/packages/console-plugin-sdk/src/typings/dashboards.ts:
```diff
@@ -77,6 +81,27 @@ namespace ExtensionProperties {
+    /**
+     * Link to all resources page.
+     * If not provided than list page of first resource from resources prop is used.
+     * */
```
fixed

rawagner commented on frontend/packages/console-shared/src/components/dashboard/status-card/OperatorStatusBody.tsx:
```diff
@@ -0,0 +1,94 @@
+  const operatorStatuses = getOperatorsWithStatuses(resources);
+  const sortedOperatorStatuses = getMostImportantStatuses(operatorStatuses)
+    .sort((a, b) => a.operators[0].metadata.name.localeCompare(b.operators[0].metadata.name))
+    .slice(0, 5);
```
yes, max 5 is by design

andybraren replied:
Hm, thinking about this again, maybe we should reconsider and show all degraded/failed operators  in the popover rather than just the first alphabetical 5 (which is arbitrary). I think that would be more useful.

rawagner replied:
Ok, limit is now removed

rawagner commented on frontend/packages/console-shared/src/components/dashboard/status-card/state-utils.ts:
```diff
@@ -0,0 +1,89 @@
+      operators: [],
+    };
+  }
+  const operatorsByStatus: { [key: string]: OperatorStatusWithResources<R> } = operators.reduce<{
```
doing
```
{} as { [key: string]: OperatorStatusWithResources<R> }
```
throws error

```
Type assertion on object literals is forbidden, use a type annotation instead.
```

vojtechszocs replied:
Is that an ESLint rule? It shouldn't do any harm on empty objects :smiley: oh well.

rawagner replied:
yes, eslint rule `eslint(@typescript-eslint/no-object-literal-type-assertion)` :)

rawagner commented on frontend/packages/console-shared/src/components/dashboard/status-card/state-utils.ts:
```diff
@@ -0,0 +1,89 @@
+    health: HealthState[statusKeys[0]],
+    detailMessage: operatorHealthPriority[statusKeys[0]].message
+      ? `${finalCount} ${operatorHealthPriority[statusKeys[0]].message}`
+      : null,
```
would you prefer not including `detailMessage` at all ? I dont think it makes any difference

vojtechszocs replied:
Yes I would prefer not including it at all, since it is marked as optional.

vojtechszocs commented:
/lgtm
